src/encoding.rs

```diff
@@ -870,23 +850,85 @@ pub mod string {
     }
 
     length_delimited!(String);
+
+    #[cfg(test)]
+    mod test {
+        use quickcheck::{quickcheck, TestResult};
+
+        use super::super::test::{check_collection_type, check_type};
+        use super::*;
+
+        quickcheck! {
+            fn check(value: String, tag: u32) -> TestResult {
+                super::test::check_type(value, tag, WireType::LengthDelimited,
+                                        encode, merge, encoded_len)
+            }
+            fn check_repeated(value: Vec<String>, tag: u32) -> TestResult {
+                super::test::check_collection_type(value, tag, WireType::LengthDelimited,
+                                                   encode_repeated, merge_repeated,
+                                                   encoded_len_repeated)
+            }
+        }
+    }
+}
+
+pub trait BytesAdapter: Default + Sized + 'static {
+    fn len(&self) -> usize;
+    fn as_slice(&self) -> &[u8];
+    fn replace(&mut self, buf: impl Buf);
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+impl BytesAdapter for Bytes {
+    fn len(&self) -> usize {
+        Buf::remaining(self)
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        Buf::bytes(self)
+    }
+
+    fn replace(&mut self, mut buf: impl Buf) {
+        // Replace bytes without allocating or copying if the underlying
+        // Buf is an instance of Bytes.
+        *self = buf.to_bytes();
+    }
+}
+
+impl BytesAdapter for Vec<u8> {
+    fn len(&self) -> usize {
+        Vec::len(self)
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        Vec::as_slice(self)
+    }
+
+    fn replace(&mut self, buf: impl Buf) {
+        // Always copies memory; and allocates if the vec has not been used
+        // before or if capacity < len;
+        self.clear();
+        self.put(buf);
+    }
 }
 
 pub mod bytes {
     use super::*;
 
-    pub fn encode<B>(tag: u32, value: &Vec<u8>, buf: &mut B)
+    pub fn encode<B>(tag: u32, value: &impl BytesAdapter, buf: &mut B)
     where
         B: BufMut,
     {
         encode_key(tag, WireType::LengthDelimited, buf);
         encode_varint(value.len() as u64, buf);
-        buf.put_slice(value);
+        buf.put_slice(value.as_slice()); // TODO Avoid copying
     }
```

Comment:
Unfortunately the `BufMut` trait does not allow zero-copy assignment, so more extensive changes are needed.
I've got [a struct](https://github.com/mzabaluev/tokio-text/blob/master/src/chunked_bytes.rs) lying around that could be used for more efficient output; it's mildly allocatey, but it could be reused for encoding successive values to save on reallocations. If you are interested, I can update it to bytes 0.5 and put it into a separate crate.

Reply:
Yeah some changes in the calling code would be necessary to support zero-copying support for encoding. This PR does not try to address that right now.

Regardless I think publishing a chain-like implementation for the latest version `bytes::BufMut` would be nice...

Reply:
Alternatively, encoding could be done entirely async with `AsyncWrite`, but this would make tokio a dependency, and I'm not sure if this project would do good to abandon synchronous (or just happily memory-chugging) use cases entirely. Perhaps there's a need for both allocating and async encoding.

Reply:
Though by the looks of `tokio_utils::codec`, Tokio developers aren't too keen on non-allocating async encoders either.

Reply:
My take is that this should be done through a rope/chain like datastructure, and not through any kind of async/CPS/state-machine style encoding.  Happy to discuss futher, but IMO outside the scope of this PR.

Reply:
> My take is that this should be done through a rope/chain like datastructure.

I have made a pre-release of `chunked-bytes`: https://github.com/mzabaluev/chunked-bytes/blob/master/examples/encoder.rs